src/plugins/kibana_utils/public/state_management/url/kbn_url_tracker.ts

```diff
@@ -0,0 +1,177 @@
+ * as long as the app is not active.
+ */
+export function createKbnUrlTracker({
+  baseUrl,
+  defaultSubUrl,
+  storageKey,
+  stateParams,
+  navLinkUpdater$,
+  toastNotifications,
+  history,
+  storage,
+}: {
+  /**
+   * Base url of the current app. This will be used as a prefix for the
+   * nav link in the side bar
+   */
+  baseUrl: string;
+  /**
+   * Default sub url for this app. If the app is currently active or no sub url is already stored in session storage and the app hasn't been visited yet, the nav link will be set to this url.
+   */
+  defaultSubUrl: string;
+  /**
+   * List of URL mapped states that should get updated even when the app is not currently active
+   */
+  stateParams: Array<{
+    /**
+     * Key of the query parameter containing the state
+     */
+    kbnUrlKey: string;
+    /**
+     * Observable providing updates to the state
+     */
+    stateUpdate$: Observable<unknown>;
+  }>;
+  /**
+   * Key used to store the current sub url in session storage. This key should only be used for one active url tracker at any given ntime.
+   */
+  storageKey: string;
+  /**
+   * App updater subject passed into the application definition to change nav link url.
+   */
+  navLinkUpdater$: BehaviorSubject<(app: AppBase) => { activeUrl?: string } | undefined>;
+  /**
+   * Toast notifications service to show toasts in error cases.
+   */
+  toastNotifications: ToastsSetup;
+  /**
+   * History object to use to track url changes. If this isn't provided, a local history instance will be created.
+   */
+  history?: History;
+  /**
+   * Storage object to use to persist currently active url. If this isn't provided, the browser wide session storage instance will be used.
+   */
+  storage?: Storage;
+}): KbnUrlTracker {
+  const historyInstance = history || createHashHistory();
+  const storageInstance = storage || sessionStorage;
+
+  // local state storing current app status and active url
+  let activeUrl: string = '';
+  let appIsMounted = false;
+
+  function setNavLink(hash: string) {
+    navLinkUpdater$.next(() => ({ activeUrl: baseUrl + hash }));
+  }
+
+  function getActiveSubUrl(url: string) {
+    // remove baseUrl prefix (just storing the sub url part)
+    return url.substr(baseUrl.length);
+  }
+
+  // track current hash when within app
+  const stopHistory = historyInstance.listen(location => {
+    if (!appIsMounted) {
+      return;
+    }
+    const urlWithHashes = baseUrl + '#' + location.pathname + location.search;
+    let urlWithStates = '';
+    try {
+      urlWithStates = unhashUrl(urlWithHashes);
+    } catch (e) {
+      toastNotifications.addDanger(e.message);
+    }
+
+    activeUrl = getActiveSubUrl(urlWithStates || urlWithHashes);
+    storageInstance.setItem(storageKey, activeUrl);
+  });
+
+  // propagate state updates when in other apps
+  const subs = stateParams.map(({ stateUpdate$, kbnUrlKey }) =>
+    stateUpdate$.subscribe(state => {
+      if (appIsMounted) {
```

Comment:
nit: there is a possible performance benefit in actually unsubscribing and resubscribing if needed, instead of using such flag. 
just an example, if `stateUpdate$` does some expensive work, e.g.:
`const stateUpdate$ = state$.pipe(filter(expensiveCheckFn))`
then, if no one is subscribed, no redundant `expensiveCheckFn` calls happens.

It is not really beneficial with specifically using current `getQueryStateContainer`, but we could improve that one later.

Reply:
Had that in briefly but it made the code more complex, that's why I went with the boolean flag  approach. You are right though, I'll give it another shot.